### DISCUSSION TITLE
How about add extra field "runtime": VM or Native?

### DISCUSSION THREAD

**Discussion by MaxGraey** [PlummersSoftwareLLC/PrimeView#16] [Ideas]
This would allow you to filter and compare languages running only in VM or only in native execution. 

For example:
|  #  | Language | ... | Runtime
|----|----|----|----|
|  1  |  C++ | ... | Native
|  2  |  Rust | ... | Native
|  3  |  Zig | ... | Native
| 4  | C# | ... | VM
| 5 | NodeJS | ... | VM
| 6 | Lua | ... | VM
| 7 | Java | ... | VM
| 8 | Dart | ... | VM       (uses DartVM)
| 9 | Dart | ... | Native  (uses Dart AOT)

**Comment by rbergen** [comment]
I'm wondering what this "runtime" field should be based on?

**Reply by MaxGraey** [reply]
I guess similar to other fields like "bits", "threads" it could be based on output string which exposed by every program? To be honest, I don't really know where these values come from. But it definitely has to be persistent.

**Reply by MaxGraey** [reply]
For C/C++
```diff
- printf("%s;%d;%0.1f;1;algorithm=wheel,faithful=yes,bits=1\n"
+ printf("%s;%d;%0.1f;1;algorithm=wheel,faithful=yes,bits=1,runtime=native\n"
```
For C#
```diff
- Console.WriteLine(
-   $"kinematics_{sieve.QuickName};\
-   {passes};{watch.Elapsed.TotalSeconds:G6};\
-   {totalThreads};algorithm={algorithm},faithful=yes{bitsSuffix}");
+ Console.WriteLine(
+   $"kinematics_{sieve.QuickName};\
+   {passes};{watch.Elapsed.TotalSeconds:G6};\
+   {totalThreads};algorithm={algorithm},faithful=yes{bitsSuffix},runtime=vm");
```

**Reply by rbergen** [reply]
We try to limit the output values to aspects that are specific to the implementations, and make those as objective as possible. With the "faithfulness" and "algorithm" tags this already poses a significant challenge.

The VM or native determination would need to be based on a combination of the language and, at least in certain cases, the execution environment the implementation chooses to use. On top of that, there are the inevitable gray areas, where one could argue either way.

To be honest, I wouldn't know how to make the VM vs Native determination objective and persistent, at least at this time. That said, I am open to suggestions on how to make this work.

**Reply by MaxGraey** [reply]
Why is this a grey area? It is always clear when it is executed natively or inside the VM (JIT, interpreted). It is a property of the compiler/runtime of the language itself. For example Dart has two variants DartVM and Dart AOT (native).

Generally it is quite easy to define this. If the target file is an executable (Linux/ELF, Windows/exe and etc) then it is probably Native, except for C# (which is not AOT).

However, all this will be checked and rechecked by the community. Just as it exists now.

**Reply by rbergen** [reply]
It's not always clear, as you yourself indicate. There are languages that support multiple environments, so it's not just a language characteristic. C# is also a good example. It is not AOT yet, but they're working on a compiler that supports it. Your use of the word "probably" also illustrates that it's not a black or white distinction.

Another issue I've already mentioned is that it's not an prime sieve implementation characteristic, which is what the output line is there to document. It's a characteristic of the runtime/execution environment. A single implementation may be executed in multiple runtimes, that may also change over time and may or may not expose their "nature" to the code they're executing. I therefore think the drag-race output line is not the place to output this.

**Reply by MaxGraey** [reply]
Then contributors should probably optionally display this in the name of the implementation, so as to somehow distinguish between that Dart AOT and DartVM. WDYT?

**Reply by rbergen** [reply]
Where would this "name of the implementation" reside?

**Reply by MaxGraey** [reply]
I mean if we add new "runtime" field for [summation algorithm string](https://github.com/PlummersSoftwareLLC/PrimeView/discussions/16#discussioncomment-1116015) in Prime Viewer we could add this column for sorting and filtering

**Reply by rbergen** [reply]
That's your original proposal, about which I've explained [here](https://github.com/PlummersSoftwareLLC/PrimeView/discussions/16#discussioncomment-1116018) and [here](https://github.com/PlummersSoftwareLLC/PrimeView/discussions/16#discussioncomment-1118560) why I don't see that as an option.

**Reply by MaxGraey** [reply]
How about this general note? https://github.com/PlummersSoftwareLLC/Primes/discussions/650#discussioncomment-1199611
